Marketplace

JivkoKelchev opened the pull request.

JivkoKelchev committed:
Added marketplace example and some TODOs

JivkoKelchev committed:
Remove all for loops

JivkoKelchev committed:
Generate bytes32 token ids

JivkoKelchev committed:
added cancelListing function

JivkoKelchev committed:
added some tests

JivkoKelchev committed:
added more tests

JivkoKelchev committed:
Refactor functions and events to be same as architecture schema. 100% coverage

vikinatora commented on contracts/LimePlace.sol:
```diff
@@ -0,0 +1,109 @@
+        bool listed;
+    }
+    
+    event LogListingAdded(bytes32 listingId, address nftContract, uint256 nftId, address seller, uint256 price);
```
The convention is to use `tokenId` instead of `nftId`.

vikinatora commented on contracts/LimePlace.sol:
```diff
@@ -0,0 +1,109 @@
+    // List the NFT on the marketplace
+    function list(address _nftContract, uint256 _tokenId, uint256 _price) public payable {
+        require(_price > 0, "Price must be at least 1 wei");
+        require(msg.value == LISTING_FEE, "Not enough ether for listing fee");
```
Asking for the fee when listing is a good approach if you want to solve the issue that appear when the fee is transferred after it's purchased - the seller might not have enough ETH to cover the fee. This approach however is not nice for the sellers. They've paid a fee but don't have a guarantee that their NFT is going to be sold and no way to get their fee back.

I'd suggest that you add a mechanism for getting this fee back if the listing isn't successfully sold.

vikinatora commented on contracts/LimePlace.sol:
```diff
@@ -0,0 +1,109 @@
+        address payable seller;
+        uint256 price;
+        bool listed;
+    }
```
The listing structure is nice. There's an advanced use case scenario that I'd like you to think about. 

You list token A for 0.1 ETH on LimePlace and OpenSea. Since OpenSea has more liquidity(this is the reality...) your NFT is purchased by a seller on OpenSea. Your listing on LimePlace is still active! After some time, you purchase token A from some marketplace for 0.5 ETH because it's undervalue and you think you can sell it for 1 ETH. 

After the purchase transaction, you expect that you will be able to see the NFT in your wallet but it's not there. You take a look at your transactions and see that a bot has executed the purchaseListing function on your old LimePlace listing for 0.1 ETH...

This scenario still happens a lot but there's something that you can do in order to prevent this case. What would you change?

vikinatora commented on contracts/LimePlace.sol:
```diff
@@ -0,0 +1,109 @@
+        
+        //check if token is supporting erc721
+        require(IERC721(_nftContract).supportsInterface(0x80ac58cd), "This marketplace support only ERC721 tokens");
+        require(IERC721(_nftContract).getApproved(_tokenId) == address(this), "LimePlace should be approved for operator");
```
I'd suggest using `isApprovedForAll` since `getApproved` allows only one contract to be able to operate the NFT. This makes is hard for people to listing on multiple marketplaces.

vikinatora commented on contracts/LimePlace.sol:
```diff
@@ -0,0 +1,109 @@
+        //check if token is supporting erc721
+        require(IERC721(_nftContract).supportsInterface(0x80ac58cd), "This marketplace support only ERC721 tokens");
+        require(IERC721(_nftContract).getApproved(_tokenId) == address(this), "LimePlace should be approved for operator");
+        _marketOwner.transfer(LISTING_FEE);
```
There's no risk of reentrancy attack here but I'd suggest making transfers(external functions) as late as possible in the function

vikinatora commented on contracts/LimePlace.sol:
```diff
@@ -0,0 +1,109 @@
+        }
+        
+        emit LogListingUpdated(_listingId, nft.nftContract, nft.tokenId, msg.sender, _price, _listed);
+    }
```
With this implementation users aren't able to cancel their listings if they're not the owner of the NFT.

vikinatora commented on contracts/LimePlace.sol:
```diff
@@ -0,0 +1,109 @@
+        _activeListings.safeRemoveToken(_listingId);
+        
+        emit LogListingSold(_listingId, nft.nftContract, nft.tokenId, nft.seller, buyer, msg.value);
+    }
```
This function is vulnerable to a reentrancy attack. Can you think of how?

vikinatora commented on contracts/LimePlace.sol:
```diff
@@ -0,0 +1,109 @@
+    
+    function generateListingId(address _contractAddress, uint256 _tokenId) public pure returns(bytes32) {
+        return keccak256(abi.encode(_contractAddress, _tokenId));
+    }
```
This will not work as soon as there're two listings for the same NFT.

vikinatora commented on contracts/LimePlace.sol:
```diff
@@ -0,0 +1,109 @@
+    mapping(bytes32 => Listing) private _listings;
+    
+    //keep track of active listings
+    TokenList.List private _activeListings;
```
Currently, you're using `_activeListings` only to fetch the currently active ones on-chain. This defeats the purpose of the API you're going to build and adds overhead to each function. The API's goal is to exactly serve you information about the listings.

JivkoKelchev committed:
Remove activeListing array.
Rename nftId->tokenId by convention.
Rearrange buy and list function to make external contract calls last.
Added Reentrancy guard.

JivkoKelchev committed:
Change approve to setApproveForAll.
Added isNotExpired modifier on buy listing function.

JivkoKelchev committed:
cancel function + fee refund on cancel

JivkoKelchev committed:
cancel function + fee refund on cancel

JivkoKelchev committed:
return listing id on list function

JivkoKelchev committed:
update tests

JivkoKelchev committed:
fix // @ts-ignore

JivkoKelchev committed:
added ownable

JivkoKelchev committed:
100% coverage

JivkoKelchev committed:
100% coverage

JivkoKelchev committed:
local deploy script

JivkoKelchev committed:
remove return listingId

JivkoKelchev committed:
up

JivkoKelchev committed:
fix expected revertWith

JivkoKelchev committed:
Merge pull request #2 from JivkoKelchev/remarks

Remarks
